service/tablet_allocator.cc

```diff
@@ -227,6 +227,105 @@ class load_balancer {
         }
     };
 
+    // We have split (target max) and merge (target min) thresholds.
+    //
+    // Naturally, merge_threshold would be X=0.5 * (split_threshold / 2), such that after merge we're half-way to split
+    // or merge. But we want to apply hysteresis to the decision, to avoid possible oscillations between split and
+    // merge. So merge_threshold will use X of 0.6, so after merge, the avg tablet size will be 60% of split_threshold,
+    // leaving 40% of room before we decide to split again. Delaying merge a bit shouldn't cause problems.
+    const uint64_t _target_max_tablet_size = default_max_tablet_size;
+    static constexpr uint64_t target_min_tablet_size(uint64_t target_max_tablet_size) {
+        return double(target_max_tablet_size / 2) * 0.6;
+    }
```

Comment:
0.5 <-> 2.0 already applies hysteresis.

If the target size is 5GB, and you have 10GB average size, you split. The new average size is 5GB and you are equally far from 2.5GB and 10GB. Similarly if you have an average size of 2.5GB and you merge, you get a new average size of 5GB again.

For this to work, you need to calculate the average tablet size using the post split/merge tablet count, even before the split/merge is complete.

Reply:
"For this to work, you need to calculate the average tablet size using the post split/merge tablet count, even before the split/merge is complete."

I don't understand completely. That's to stabilize load balancer decisions?

If we have 8 tablets, with avg of 10G. Then we split.

It's still important to calculate new averages with 8 tablet count, in order to allow cancellation, if avg drops below 5G before split completes.

If during split, we say tablet count == 16, then avg would be 5G immediately, misleading the balancer. In my current understanding, the balancer should keep working with the real tablet count, regardless of any ongoing resize decision.

Reply:
No, you should calculate the new averages with 16 tablets. Then the average jumps to 5G, and you'll make a cancellation decision when it drops to 2.5G. Then you don't need the 0.6 magic number.

Reply:
> If during split, we say tablet count == 16, then avg would be 5G immediately, misleading the balancer. In my current understanding, the balancer should keep working with the real tablet count, regardless of any ongoing resize decision.

Why does it mislead the load balancer?

The load balancer can still consider unsplit tablets, but to decide whether to cancel or not, it can use the post-split/merge sizes. It's saying "if this split continues, will I need to merge?"